source/docs/software/vision-processing/apriltag/apriltag-intro.rst

```diff
@@ -0,0 +1,233 @@
+
+      * Calculate the expected interior pixel coordinates where the center of each bit should be
+      * Mark each location as "1" or "0" by comparing the pixel intensity to a threshold
+      * Find the tag ID which most closely matches what was seen in the image, allowing for one or two bit errors.
+
+      It is possible there is no valid tag ID which matches the suspect tag. In this case, the decoding process stops.
+
+   .. tab:: Fit External Quad
+
+      .. image:: images/fit_ext_quad.png
+
+      Now that we have a tag ID for the region of pixels, we need to do something useful with it.
+
+      For most FRC applications, we care about knowing the precise location of the corners of the tag, or its center. In both cases, we expect the resolution-lowering operation we did at the beginning to have distorted the image, and we want to undo those effects.
+
+      The algorithm to do this is:
+
+      * Use the detected tag location to define a region of interest in the original-resolution image
+      * Calculate the :term:`gradient` at pre-defined points in the region of interest to detect where the image most sharply transitions between black to white
+      * Use these gradient measurements to rapidly re-fit an exterior quadrilateral at full resolution
+      * Use geometry to calculate the exact center of the re-fit quadrilateral
+
+      Note that this step is optional, and can be skipped for faster image processing. However, skipping it can induce significant errors into your robot's behavior, depending on how you are using the tag outputs.
+
+
+Usage
+-----
+
+2D Alignment
+^^^^^^^^^^^^
+
+A simple strategy for using targets is to move the robot until the target is centered in the image. Assuming the field and robot are constructed such that the gamepiece, scoring location, vision target, and camera are all aligned, this method should proved a straightforward method to automatically align the robot to the scoring position.
+
+Using a camera, identify the _centroid_ of the apriltags in view. If the tag's ID is correct, apply drivetrain commands to rotate the robot left or right until the tag is centered in the camera image.
+
+This method does not require calibrating the camera or performing the homography step.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    // Coming soon!
+
+  .. code-tab:: c++
+
+    // Coming soon!
+
+3D Alignment
+^^^^^^^^^^^^
+
+.. image:: images/homography.png
+
+A more advanced usage of AprilTags is to use their corner locations to help perform on-field localization.
+
+This method requires calibrating the camera to measure how its optics distort 3d space onto the 2d sensor.
+
+Once calibrated, each image is searched for AprilTags using the above algorithm.
+
+Given each tag's ID, the position of the tag on the field can be identified.
```

Comment:
Can this be made a paragraph? Maybe it's just me, but reading sentences segmented is a bit jarring.